Code to convert WBS table to BLT

ianbjacobs opened the pull request:
Generate BLT from a W3C WBS election results table using ranked voting controls. See code comments for assumptions about valid and invalid ballots.

ianbjacobs committed:
V0.9 for review

dontcallmedom commented on table-to-blt.mjs:
```diff
@@ -0,0 +1,147 @@
+
+async function main(file, nbseats, electionname) {
+    const html = await readFile(file, { encoding: 'utf8'});
+    let dom = new JSDOM(html);
```
you could remove the dependency to `node:fs` and simplify these two lines with
```js
const dom = await JSDOM.fromFile(file);
```
(`const` since you don't expect to re-assign that variable - likely applies to all the other `let`s)

dontcallmedom commented on table-to-blt.mjs:
```diff
@@ -0,0 +1,147 @@
+  let table = dom.window.document.querySelector("table")
+
+  // Ignore first column in determining number of candidates.
+  const nbcandidates = table.querySelector("tr").querySelectorAll('td,th').length - 1;
```
alternatively, since the first column *is* the `th`:
```suggestion
  const nbcandidates = table.querySelector("tr").querySelectorAll('td').length;
```

dontcallmedom commented on table-to-blt.mjs:
```diff
@@ -0,0 +1,147 @@
+  let rows = table.querySelectorAll("tr");
+
+  // candidates is the list of candidate names, taken from the top (zeroth) row of the table.
+  const candidates = Array.from(rows[0].querySelectorAll('th')).map(s => s.innerHTML).slice(1) ;
```
alternative syntax sugar:
```suggestion
  const candidates = [...rows[0].querySelectorAll('th')].map(s => s.innerHTML).slice(1) ;
```
Also, you're probably not interested in `innerHTML` but instead `textContent` (here and in other usages later in the code).

dontcallmedom commented on table-to-blt.mjs:
```diff
@@ -0,0 +1,147 @@
+    generateBLT(dom, nbseats, electionname);
+}
+
+function generateBLT (dom, nbseats, electionname) {
```
most of this code could be re-used easily directly in the browser (e.g. as a bookmarklet or via the developer tools console) if it used `document` (from `dom.window.document`) rather than `dom` as the source of the content being passed across functions)

dontcallmedom commented on table-to-blt.mjs:
```diff
@@ -0,0 +1,147 @@
+  
+  // Start output in BLT format
+  // First row is number of candidates and number of seats
+  console.log(nbcandidates + " " + nbseats);
```
modern JS alternative to string concatenation
```suggestion
  console.log(`${nbcandidates} ${nbseats}`);
```

dontcallmedom commented on table-to-blt.mjs:
```diff
@@ -0,0 +1,147 @@
+  console.log("0");
+  // Names of candidates, ignoring the first column.
+  for (let i = 1; i <= nbcandidates; i++) {
+    console.log("\"Candidate " + i + "\"");
```
(purely a personal preference - I think I'd find it less footgunny to have a regular `for` loop (`let i = 0; i< nbcandidates`…) and use `i+1` in the `console.log`)

dontcallmedom commented on table-to-blt.mjs:
```diff
@@ -0,0 +1,147 @@
+	}
+
+        // Sort the candidates so the array of cells goes from top ranked (1) to lowest ranked.
+        let rankedcells = [...unranked].sort(sortByRank);
```
I assume you're using `[...unranked]` to duplicate the array and leave `unranked` alone; there has been a very recent addition to JS that does this directly (but only starting node 20 and latest versions of browsers):
```suggestion
        const rankedcells = unranked.toSorted(sortByRank);
```

Separately, since `sortByRank` is only used once, you could inline it
```js
  const rankedcells = unranked.toSorted((a,b) => a.rank - b.rank);
```

dontcallmedom commented on table-to-blt.mjs:
```diff
@@ -0,0 +1,147 @@
+        // Sort the candidates so the array of cells goes from top ranked (1) to lowest ranked.
+        let rankedcells = [...unranked].sort(sortByRank);
+
+        // Per OpenSTV, ballot is invalid if there are skips in rankings.
```
the alternative (which I think you had initially) would be to fix the ballot by skipping the gaps; since that's at the intersection of UI (should WBS allow invalid ballots?) & election integrity, that may be worth at least some explicit discussion with the election organizers.

dontcallmedom commented on table-to-blt.mjs:
```diff
@@ -0,0 +1,147 @@
+    // Handle rows. Ignore top (zeroth) row since that is candidate names.
+    for (let i = 1; i < rows.length; i++) {
+        // Handle cells in the row.
+        let cells = rows[i].querySelectorAll('td');
```
when walking an array, I would usually use
```suggestion
        for (const row of rows) {
```
IOW, `i` is a distraction when the real focus of the loop is `rows[i]` - (here, and in other similar loops).

In this case, since you're skipping the first row, it would be 
```suggestion
        for (const row of rows.slice(1)) {
```

dontcallmedom commented on table-to-blt.mjs:
```diff
@@ -0,0 +1,147 @@
+        let rankedcells = [...unranked].sort(sortByRank);
+
+        // Per OpenSTV, ballot is invalid if there are skips in rankings.
+        if (!noSkips(rankedcells, cells.length)) {
```
isn't `cells.length` necessarily equal to `rankedcells.length` at this stage? if so, the second argument doesn't seem useful, it can be derived in the function itself.

dontcallmedom commented on table-to-blt.mjs:
```diff
@@ -0,0 +1,147 @@
+        for (let j = 0; j < cells.length; j++) {
+	  // Unranked candidates are demoted in the list by giving them very high rank (Number.POSITIVE_INFINITY).
+	  // Remove these candidates from the generated ballot.
+	  if (rankedcells[j].rank != Number.POSITIVE_INFINITY) {
```
```suggestion
	  if (rankedcells[j].rank !== Infinity) {
```

dontcallmedom commented on table-to-blt.mjs:
```diff
@@ -0,0 +1,147 @@
+  const re2 = /\s*Unranked\s*/ ;
+  let title = cell.getAttribute('title');
+  if (title != null) {
+     return { candidate: title.split(':')[0], rank: title.split(':')[1].replace(re1, "$1").replace(re2, Number.POSITIVE_INFINITY) };
```
This generates a string value for rank, when we really mean to get a number.

I'd go with something like:
```js
const [ candidate, strRank ] = title.split(':');
const rank = strRank.match(re2) ? Infinity : parseInt(strRank.replace(re1, "$1"), 10);
return { candidate, rank };
```

dontcallmedom commented on table-to-blt.mjs:
```diff
@@ -0,0 +1,147 @@
+     return { candidate: title.split(':')[0], rank: title.split(':')[1].replace(re1, "$1").replace(re2, Number.POSITIVE_INFINITY) };
+
+  } else {
+     return ("");
```
if this situation arises, this would make other code breaks (e.g. when trying to sortByRank); since I think we don't expect that situation to be valid, I would throw an exception
```js
throw new Error(`Cell ${cell.outerHTML} does not follow expected format`);
```

ianbjacobs committed:
Changes based on Dom's review

ianbjacobs commented:
Thanks, Dom! I took into account all of your suggestions. I also added some additional error checking. I also tested on some (non-public) election data locally.

Regarding your point about throwing away ballots with skips, I added a parameter to allow them. (By default they are considered invalid, however.)

dontcallmedom commented on table-to-blt.mjs:
```diff
@@ -29,76 +29,78 @@
-  for (let i = 1; i <= nbcandidates; i++) {
-    console.log("\"Candidate " + i + "\"");
+  for (let i = 0; i < nbcandidates; i++) {
+    console.log(`\"Candidate ${i + 1}\"`);    
```
you don't need to escape quotes (here and elsewhere)

dontcallmedom commented on table-to-blt.mjs:
```diff
@@ -29,76 +29,78 @@
         // Per OpenSTV, ballot is invalid if there are skips in rankings.
-        if (!noSkips(rankedcells, cells.length)) {
-	   console.error("Ballot ignored (skips in rankings): " + rows[i].querySelector('th').innerHTML);
+	// However, if ignoreskips is true, allow ballot with skips.
```
I think this would need more than not ignoring them - we would want to fix them by removing gaps IIUC (otherwise they'll get rejected by the tool)

ianbjacobs committed:
editorial: escaping quotes unnecessary

ianbjacobs committed:
Remove ignore skips boolean since more thought is required.

ianbjacobs commented:
@dontcallmedom,

I've removed the ignoreskips bits for now. More consideration will be needed about whether and how to handle duplicate votes, skips, and withdrawn candidates. I suggest we merge this version with the expectation it will be used with OpenSTV, then we enhance to work with OpaVote once we know how we want to handle those scenarios.

ianbjacobs committed:
simpler skip tests

ianbjacobs committed:
Shownames option to show (v anonymize) names

ianbjacobs committed:
typo fix

ianbjacobs committed:
Like this duplicate test better

ianbjacobs committed:
cleaned up a comment.

ianbjacobs committed:
Refinements

ianbjacobs committed:
clean up comments a bit

ianbjacobs committed:
Fixed test

ianbjacobs committed:
Add expected output
